common/emplace_result.h

```diff
@@ -0,0 +1,48 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_EMPLACE_RESULT_H_
+#define CARBON_COMMON_EMPLACE_RESULT_H_
+
+#include <type_traits>
+#include <utility>
+
+namespace Carbon {
+
+// A utility to use when calling an `emplace` function to emplace the result of
```

Comment:
I am thoroughly confused how this is changing the behaviour of emplace, could you explain it a bit here?

Reply:
Extended the comment to explain why this works.